### PR TITLE
Add road types: triangle and bump

### PR DESCRIPTION
**Pull request by clemens-tolboom**
I wanted to see what would happen on a triangular road and semi triangular (bump + flat)

### Triagle

triagle

### Bump

bump

### CODE REVIEW

File: example.py [hunk 1]
```diff
@@ -44,6 +44,9 @@ def simulate(car, time_step=0.0002, interval=1):
 
 if __name__ == "__main__":
     car = Car()
+    # road = Road(length= 6, mode="triagle", frequency=1.8, amplitude= 0.1)
```
Comment (clemens-tolboom) [inline example.py]: This needs x_min like below

```python
    road = Road(length= 6, x_min=-3, mode="triagle", frequency=1.8, amplitude= 0.1)
    car = Car(road_func=road)
```

File: halfcar/road.py [hunk 2]
```diff
@@ -84,15 +84,29 @@ def generate(self, length_to_generate):
         for i in range(num_new_points):
             self.y_coords.popleft()
 
-            if self.mode in ("sine", "square"):
-                sine_value = math.sin(frequency * (distance + (i / resolution)))
+            if self.mode in ("sine", "square", "triagle", "bump"):
```
Comment (nrsyed) [inline halfcar/road.py]: Fixing minor typo.

```suggestion
            if self.mode in ("sine", "square", "triangle", "bump"):
```

File: halfcar/road.py [hunk 3]
```diff
@@ -84,15 +84,29 @@ def generate(self, length_to_generate):
...
+                sin_arg = frequency * (distance + (i / resolution))
+                sine_value = math.sin(sin_arg)
                 if self.mode == "sine":
                     next_point = amplitude * sine_value
                 elif self.mode == "square":
                     if sine_value >= 0:
                         next_point = 0
                     else:
                         next_point = amplitude
+                elif self.mode in ("triagle", "bump"):
```
Comment (nrsyed) [inline halfcar/road.py]: Fixing minor typo.

```suggestion
                elif self.mode in ("triangle", "bump"):
```

File: example.py (hunk 1 above)
Comment (nrsyed) [reply example.py]: Please see comment on PR #4 (https://github.com/nrsyed/half-car/pull/4#discussion_r310421270) on default and hard-coded `Road` `length` and `x_min` values.

**Review by nrsyed — COMMENTED** [review]
Very nice, I like the triangle and bump road profiles. Please see comments for issues to discuss/address before ready for merging.

File: halfcar/road.py [hunk 4]
```diff
@@ -84,15 +84,29 @@ def generate(self, length_to_generate):
...
+                sine_value = math.sin(sin_arg)
                 if self.mode == "sine":
                     next_point = amplitude * sine_value
                 elif self.mode == "square":
                     if sine_value >= 0:
                         next_point = 0
                     else:
                         next_point = amplitude
+                elif self.mode in ("triangle", "bump"):
+                    # https://en.wikipedia.org/wiki/Sawtooth_wave
```
Comment (clemens-tolboom) [inline halfcar/road.py]: This ref is not relevant.

**Comment by clemens-tolboom** [comment]
I still have trouble with fixing #4. I'll try to fix this week.

File: halfcar/road.py (hunk 3 above)
Comment (clemens-tolboom) [reply halfcar/road.py]: Thanks

File: halfcar/road.py (hunk 4 above)
Comment (clemens-tolboom) [reply halfcar/road.py]: Removed

File: halfcar/road.py (hunk 2 above)
Comment (clemens-tolboom) [reply halfcar/road.py]: Thanks

File: example.py (hunk 1 above)
Comment (clemens-tolboom) [reply example.py]: I closed #4 in favour of this #5 as it is needed here to make this PR work.

**Comment by clemens-tolboom** [comment]
I can't stop watching the car taking bumps. Too much pressure on the springs :-)

**Comment by nrsyed** [comment]
> I can't stop watching the car taking bumps. Too much pressure on the springs :-)

Ha, it is nice, isn't it?

**Comment by clemens-tolboom** [comment]
Thanks!

**Pull request merged**